### ISSUE TITLE
Difficulty in getting the plugin to play nice

### ISSUE THREAD

**Issue by alexgorbatchev** [Microsoft/TypeScript-Sublime-Plugin#384]
I'm having difficulty getting `tsc` and the plugin to behave the same. I've configured the project where running `webpack` with `ts-loader` compiles and everything works perfectly using `tsconfig.json` without using `reference` tags. My `tsconfig.json` is pretty standard:

```
{
  "compilerOptions": {
    "declaration": false,
    "experimentalDecorators": true,
    "jsx": "react",
    "module": "commonjs",
    "moduleResolution": "node",
    "noEmit": true,
    "noEmitOnError": true,
    "noLib": false,
    "preserveConstEnums": true,
    "removeComments": false,
    "sourceMap": true,
    "target": "es5"
  },
  "exclude": [
    "node_modules"
  ],
  "files": [
    "typings/tsd.d.ts"
  ]
}
```

With this configuration the Sublime Plugin is basically useless as it doesn't seem to pickup on my definitions. At this time I found two options to get the plugin to work:

1. add files to the `files` node
2. add `reference` tags to all the files

I'm trying to avoid adding `reference` tags as much as possible. When I add files to the `files` node the sublime plugin works, but typescript compilation now fails with errors about duplicate variable declarations.

Does anyone have an example of a project that works fine in sublime?

**Comment by zhengbli** [comment]
Can you elaborate what you mean by "not picking up your definitions"? Do you mean that the plugin included files were not in the project?

**Comment by alexgorbatchev** [comment]
I mean even though I have all required definitions for node and other modules, the plugin still tells me that things like `require` and `__dirname` are not declared and and `import * as path from 'path'` can't find module `path`. However the project compiles just fine via `webpack`.

**Comment by alexgorbatchev** [comment]
Here are a few oddities from the https://github.com/alexgorbatchev/gulp-typescript-webpack-react-hotreload/commit/21c915c4e0b7d77d31679d27d6e1f5c472c4b4ac

![component tsx hot-reload-example_20151026204503](https://cloud.githubusercontent.com/assets/65633/10748967/89e9be02-7c22-11e5-8071-e859291d4242.png)

however in a different file the same import statement is OK

![index tsx hot-reload-example_20151026204614](https://cloud.githubusercontent.com/assets/65633/10748981/a02e241e-7c22-11e5-8e0b-52a5de4fdb89.png)

and this file seems to work fine without `reference` tag all together

![hello-world test ts hot-reload-example_20151026204641](https://cloud.githubusercontent.com/assets/65633/10748992/b8e8a574-7c22-11e5-80c2-ec7c7053a263.png)

At the same time `gulp webpack` compiles the project and `karma start` works as expected.

Would really appreciate any pointers...

**Comment by billti** [comment]
If you specify the "files" property, then ONLY those files get given to the compiler. It will not include any other files in that context unless referenced. Seeing as you only have "typings/tsd.d.ts" in your "files" array, then any other file is opened as a standalone file.

You'll need to either list all your files for the project in tsconfig.json, or put is somewhere where you want all files picked up underneath and not include a "files" list. I couldn't say why you are getting duplicate identifiers.

**Comment by alexgorbatchev** [comment]
I figured out the issue with the module and double declaration. If the file has `module.exports` in it, it's not treated as a separate module for some reason but more of a concatenation case. When I replaced `module.exports` with `exports default` it all started working fine.

Bottom line, using `gulp` plugin to keep `files` node in sync with the file system and ES5/6 syntax instead of commonjs makes it all work.
